File: 70.climbing-stairs.ts
/*
 * @lc app=leetcode id=70 lang=typescript
 *
 * [70] Climbing Stairs
 */

// @lc code=start
function climbStairs(n: number): number {
  let sum = 1;
  let prevSum = 1;
  for (let i = 0; i < n - 1; i++) {
    let temp = sum;
    sum = sum + prevSum;
    prevSum = temp;
  }
  return sum;
}
// @lc code=end

File: 70.climbing-stairs.py
#
# @lc app=leetcode id=70 lang=python3
#
# [70] Climbing Stairs
#

# @lc code=start
# dp solution
""" class Solution:
    def climbStairs(self, n: int) -> int:
        a, b = 1, 1
        for _ in range(n):
            a, b = b, a+b
        return a
 """

# recursive with memoization solution


class Solution:
    def climbStairs(self, n: int, memo={1: 1, 2: 2, 3: 3}) -> int:
        if n in memo:
            return memo[n]
        memo[n] = self.climbStairs(n-1)+self.climbStairs(n-2)
        return memo[n]
# @lc code=end
